source/opt/dataflow.h

```diff
@@ -0,0 +1,170 @@
+  // worklist.
+  bool Enqueue(Instruction* inst);
+
+ private:
+  // Runs one pass, calling |InitializeWorklist| and then iterating through the
+  // worklist until all fixed.
+  VisitResult RunOnce(Function* function, bool is_first_iteration);
+
+  friend class WorklistComparator;
+  class WorklistComparator {
+    DataFlowAnalysis& analysis_;
+
+   public:
+    WorklistComparator(DataFlowAnalysis& analysis) : analysis_(analysis) {}
+
+    bool operator()(Instruction* first, Instruction* second) {
+      return analysis_.InstructionShouldBeVisitedAfter(first, second);
+    }
+  };
+
+ protected:
+  std::unordered_map<Instruction*, bool> on_worklist_;
+  std::priority_queue<Instruction*, std::vector<Instruction*>,
+                      WorklistComparator>
+      worklist_;
+  IRContext& context_;
+};
+
+// A generic data flow analysis, specialized for forward analysis.
+class ForwardDataFlowAnalysis : public DataFlowAnalysis {
+ public:
+  // Indicates where labels should be in the worklist RPO ordering.
+  enum class LabelPosition {
+    // Labels should be placed at the beginning of their blocks.
+    kLabelsAtBeginning,
+    // Labels should be placed at the end of their blocks.
+    kLabelsAtEnd,
+    // Labels should not be in the worklist.
+    kNoLabels,
+    // Only labels should be placed in the worklist.
+    kLabelsOnly,
+  };
+
+  ForwardDataFlowAnalysis(IRContext& context, LabelPosition label_position)
+      : DataFlowAnalysis(context),
+        label_position_(label_position),
+        rpo_indices_() {}
+
+ protected:
+  // Initializes the worklist in reverse postorder, regardless of
+  // |is_first_iteration|. Labels are placed according to the label position
+  // specified in the constructor.
+  void InitializeWorklist(Function* function, bool is_first_iteration) override;
+
+  // Enqueues the users and block successors of the given instruction.
+  // See |EnqueueUsers| and |EnqueueBlockSuccessors|.
+  void EnqueueSuccessors(Instruction* inst) override {
+    EnqueueUsers(inst);
+    EnqueueBlockSuccessors(inst);
+  }
+
+  // Enqueues the users of the given instruction.
+  void EnqueueUsers(Instruction* inst);
+
+  // Enqueues the labels of the successors of the block corresponding to the
+  // given label instruction. Does nothing for other instructions.
+  void EnqueueBlockSuccessors(Instruction* inst);
+
+  // Compares the two instructions by reverse post-order.
+  // Unknown instructions are placed at the end of the ordering and are
+  // unordered within themselves (i.e. !(x > y || y > x) for all x, y not seen
+  // in |InitializeWorklist|.)
+  bool InstructionShouldBeVisitedAfter(Instruction* first,
+                                       Instruction* second) override;
+
+ private:
+  LabelPosition label_position_;
+
+ protected:
+  // Map from instruction to index in RPO ordering, starting from 0.
+  std::unordered_map<Instruction*, uint32_t> rpo_indices_;
```

Comment:
Same us in the base class: why is this exposes?

Reply:
I believe we need to expose this, since if a derived class wants to change the RPO order (for example by inserting custom instructions), they would need to modify `rpo_indices_`. What do you think?

Reply:
I'd prefer an enqueue-like method, if it does not add to much code. And an accessor for checking the order.

Reply:
Done; please take a look.